File: Day 06/src/Part_1.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Part_1 {

	public static final Pattern PATTERN = Pattern.compile("(.*?) ([0-9]+,[0-9]+) through ([0-9]+,[0-9]+)");

	public static void main(String[] args) {
		try {
			Boolean[][] states = new Boolean[1000][];
			for (int x = 0; x < 1000; x++) {
				states[x] = new Boolean[1000];
				for (int y = 0; y < 1000; y++)
					states[x][y] = false;
			}
			BufferedReader reader = new BufferedReader(new FileReader(new File("input")));
			String line;
			while ((line = reader.readLine()) != null) {
				Matcher matcher = PATTERN.matcher(line);
				matcher.find();
				String instr = matcher.group(1);
				String[] pos1 = matcher.group(2).split(",");
				String[] pos2 = matcher.group(3).split(",");
				int[] min = new int[]{ Integer.parseInt(pos1[0]), Integer.parseInt(pos1[1]) };
				int[] max = new int[]{ Integer.parseInt(pos2[0]), Integer.parseInt(pos2[1]) };
				switch (instr) {
					case "toggle":
						for (int x = min[0]; x <= max[0]; x++)
							for (int y = min[1]; y <= max[1]; y++)
								states[x][y] = !states[x][y];
						break;
					case "turn on":
						for (int x = min[0]; x <= max[0]; x++)
							for (int y = min[1]; y <= max[1]; y++)
								states[x][y] = true;
						break;
					case "turn off":
						for (int x = min[0]; x <= max[0]; x++)
							for (int y = min[1]; y <= max[1]; y++)
								states[x][y] = false;
						break;
				}
			}
			reader.close();
			int lights = 0;
			for (int x = 0; x < 1000; x++)
				for (int y = 0; y < 1000; y++)
					if (states[x][y])
						lights++;
			System.out.format("There are %d lights on.\n", lights);
		} catch (NumberFormatException | IOException e) {
			e.printStackTrace();
		}
	}

}

File: Day 8/Part 1.kt
package xyz.nickr

val pattern = java.util.regex.Pattern.compile("\\\\x([0-9A-Fa-f]{2})")

fun main(args : Array<String>) {
    val file = java.io.File("input")
    val reader = file.bufferedReader()
    var line = reader.readLine()
    var clens = 0
    var alens = 0
    while (line != null) {

        val len = line.length
        var repl = line.substring(1, len-1)
            repl = repl.replace("\\\"", "\"")
            repl = repl.replace("\\\\", "\\")

        val matcher = pattern.matcher(repl)
        var offset  = 0

        while (matcher.find()) {
            var ch = java.lang.Integer.parseInt(matcher.group(1), 16).toChar()
            repl = repl.substring(0, matcher.start() - offset) + ch +
                   repl.substring(matcher.end() - offset)
            offset += matcher.group(1).length + 1
        }
        
        clens += len
        alens += repl.length

        line = reader.readLine()

    }
    println(clens - alens)
}